### PR TITLE
feature: new endpoints for communities

### LINKED ISSUE #58: New API endpoints for communities
- [x] `GET /users/profile/fid/{fid}` | `/users/profile/username/{username}`: Get Farcaster user by FID or username.
- [x] `GET /channels?q={needle}`: Query about Warpcast channels.
- [x] `GET /channels/{channelID}`: Information about Warpcast channel, if it exists.
- [x] Validate default census:
  - [x] `POST /census/exists/[nft/erc20]`: If it is a contract, validate that it exists in the associated network.
  `Body: {"address": "", "blockchain": "base"}`
  - [x] `POST /census/exists/channel-gated`: If it is a channel, validate that it exists.
  `Body: {"channelID": "vote"}`. **replaced by existing** `GET /channels/{channelID}`
- [x] Community related endpoints:
  - [x] `GET /communities`: List communities, accepts params `byAdminFID=fid` & `byAdminUsername=username` to filter by the community admin.
  - [x] `GET /communities/{communityID}`: Get community information.
  - [x] `POST /census/community`: Create a census for a existing community.
    `Body: { "communityID": <communityID> }`

### PR DESCRIPTION
**Pull request by lucasmenendez**
Closes #58

### CODE REVIEW

File: census.go [hunk 1]
```diff
@@ -468,6 +468,147 @@ func (v *vocdoniHandler) censusFollowers(msg *apirest.APIdata, ctx *httprouter.H
...
+		data, err := json.Marshal(map[string]string{"censusId": censusID.String()})
+		if err != nil {
+			return err
+		}
+		return ctx.Send(data, http.StatusOK)
+	}
+	// if the census type is not a channel, the type is NFT or ERC20, so create
+	// the census sync and from the token holders from airstack
+
+	// extract userFID from auth token
```
Comment (p4u) [inline census.go:572]: This should be done at the beginning of the method and we should check if the user calling the endpoint is actually an admin for the community.

File: census.go [hunk 2]
```diff
@@ -468,6 +468,147 @@ func (v *vocdoniHandler) censusFollowers(msg *apirest.APIdata, ctx *httprouter.H
...
+		return ctx.Send([]byte("community not found"), http.StatusNotFound)
+	}
+	// create a censusID for the queue and store into it
+	censusID, err := v.cli.NewCensus(api.CensusTypeWeighted)
+	if err != nil {
+		return err
+	}
+	// check the type to create it from the correct source (channel or airstak
+	// (nft/erc20)) and in the correct way (async or sync)
+	if community.Census.Type == mongo.TypeCommunityCensusChannel {
```
Comment (p4u) [inline census.go:501]: Please move this code to a separate function and try to reuse the one already existing for creating channel censuses.

**Review by p4u — CHANGES_REQUESTED** [review]

File: census.go [hunk 3]
```diff
@@ -468,6 +417,89 @@ func (v *vocdoniHandler) censusFollowers(msg *apirest.APIdata, ctx *httprouter.H
...
+			Address:    addr.Address,
+			Blockchain: addr.Blockchain,
+		})
+	}
+	// check valid token
+	if err := v.checkTokens(censusAddresses); err != nil {
+		return err
+	}
+	// convert the census type to the correct type for the CreateCensus function
+	censusType := ERC20type
```
Comment (p4u) [inline census.go:480]: Nit: a switch case works better here and it's more extensible to other censuses in the future

File: census.go [hunk 4]
```diff
@@ -468,6 +417,89 @@ func (v *vocdoniHandler) censusFollowers(msg *apirest.APIdata, ctx *httprouter.H
...
+// community can be of type channel, NFT, or ERC20. If the community is a
+// channel, the census is created from the users who follow the channel, and
+// the process is async. If the community is an NFT or ERC20, the census is
+// created from the token holders of the token addresses in the community, using
+// the AirStack API. The process is sync and the census is created in the same
+// request. The census is created from the participants and the progress is
+// updated in the queue.
+func (v *vocdoniHandler) censusCommunity(msg *apirest.APIdata, ctx *httprouter.HTTPContext) error {
+	// extract userFID from auth token
+	userFID, err := v.db.UserFromAuthToken(msg.AuthToken)
```
Comment (p4u) [inline census.go:430]: We should check if the userFID is allowed to create census for such community

File: census.go [hunk 5]
```diff
@@ -468,6 +417,89 @@ func (v *vocdoniHandler) censusFollowers(msg *apirest.APIdata, ctx *httprouter.H
...
+		}
+		tokenDecimals, err = v.airstack.TokenDecimalsByToken(censusAddresses[0].Address, censusAddresses[0].Blockchain)
+		if err != nil {
+			return fmt.Errorf("cannot get token decimals")
+		}
+	}
+	// create the census from the token holders
+	data, err := v.censusTokenAirstack(censusAddresses, censusType, tokenDecimals, userFID)
+	if err != nil {
+		return fmt.Errorf("cannot create erc20 census: %w", err)
```
Comment (p4u) [inline census.go:498]: error message is wrong, its not always 'erc20 census'

File: communities.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+				FID:         user.UserID,
+				Username:    user.Username,
+				DisplayName: user.Displayname,
+				// Avatar:      user.Avatar,
+			})
+		}
+		// get census channel or addresses based on the type
+		cAddresses, cChannel, err := v.censusChannelOrAddresses(ctx.Request.Context(), c.Census)
+		if err != nil {
+			if err == farcasterapi.ErrChannelNotFound {
```
Comment (p4u) [inline communities.go:111]: This is weird. Why are you just wrapping the ErrChannelNotFound error (returning 404) while other error are 500? Both errors should be wrapper or none. I'd say not to wrap the error and just return whatever the v.censusChannelOrAddresses() method returns with a 404.

File: communities.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+	if dbCommunity == nil {
+		return ctx.Send([]byte("Community not found"), http.StatusNotFound)
+	}
+	// get admin profiles for the community
+	admins := []*User{}
+	for _, admin := range dbCommunity.Admins {
+		user, err := v.db.User(admin)
+		if err != nil {
+			if err == farcasterapi.ErrNoDataFound {
+				continue
```
Comment (p4u) [inline communities.go:161]: Do not silent the errors please, it's ok to keep going but print it at least as a Warning on the logs.

**Review by p4u — CHANGES_REQUESTED** [review]

File: farcasterapi/neynar/neynar.go [hunk 8]
```diff
@@ -422,18 +440,48 @@ func (n *NeynarAPI) ChannelFIDs(ctx context.Context, channelID string, progress
...
 		}
-		return false, fmt.Errorf("error creating request: %w", err)
+		return false, fmt.Errorf("error checking channel existence: %w", err)
 	}
 	return true, nil
 }
 
+// FindChannel method returns the list of channels that match the query provided.
+// If something goes wrong, it returns an error.
+func (n *NeynarAPI) FindChannel(ctx context.Context, query string) ([]*farcasterapi.Channel, error) {
```
[the remark this answers is not in the document]
Comment (lucasmenendez) [reply farcasterapi/neynar/neynar.go:456]: There is no alternative to neynar to get information about channels. They are not in the farcaster protocol, internally they are like a thread of a parent cast (the channel itself), and their properties are not registered in the hubs (such as followers count or followers list).

[the remark this answers is not in the document]
Comment (lucasmenendez) [reply farcasterapi/neynar/neynar.go:456]: Discover and track channels requires to create a service to scan internally using the Neynar API.

[the remark this answers is not in the document]
Comment (p4u) [reply farcasterapi/neynar/neynar.go:456]: Still wondering what happens if the query is empty. And wondering also if we should handle this situation.

**Review by p4u — APPROVED** [review]

**Pull request merged**